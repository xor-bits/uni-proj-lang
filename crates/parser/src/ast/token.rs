use lexer::Token;
use macros::Parse;

use crate::{Parse, ParseStream, Result, SingleToken};

#[cfg(test)]
use serde::Serialize;

//

macro_rules! simple_tokens {
    ($($(#[$($t:tt)*])? $variant:ident),* $(,)?) => {$(
        $(#[$($t)*])?
        #[cfg_attr(test, derive(Serialize))]
        #[derive(Debug, Clone, Copy, PartialEq, Eq, Parse)]
        #[token(Token::$variant)]
        pub struct $variant;
    )*};
}

// #[macro_export]
// #[rustfmt::skip]
// macro_rules! Tok {
//     (;) => { ast::token::Semi };
//     (=) => { ast::token::Assign };
//     // (() => { ast::token::};
//     // ()) => { ast::token::};
//     // ({) => { ast::token::};
//     // (}) => { ast::token::};
//     (+) => { ast::token::Plus };
//     (-) => { ast::token::Minus };
//     (*) => { ast::token::Asterisk };
//     (/) => { ast::token::Slash };
//     (:) => { ast::token::Colon };
//     (,) => { ast::token::Comma };
//     (@) => { ast::token::At };
//     (&) => { ast::token::Ampersand };
//     (<) => { ast::token::Lt };
//     (>) => { ast::token::Gt };
//     (<=) => { ast::token::Le };
//     (>=) => { ast::token::Ge };
//     (==) => { ast::token::Eq };
//     (!=) => { ast::token::Neq };
//     (:=) => { ast::token::Walrus };
//     (->) => { ast::token::RArrow };
//     (struct) => { ast::token::Struct };
//     (test) => { ast::token::Test };
// }

// fn x() {
//     let x: Tok![struct] = todo!();
// }

//

simple_tokens! {
    /// `;`
    Semi,

    /// `=`
    Assign,

    /// `(`
    LParen,

    /// `)`
    RParen,

    /// `{`
    LBrace,

    /// `}`
    RBrace,

    /// `+`
    Plus,

    /// `-`
    Minus,

    /// `*`
    Asterisk,

    /// `/`
    Slash,

    /// `:`
    Colon,

    /// `,`
    Comma,

    /// `@`
    At,

    /// `&`
    Ampersand,

    /// `<`
    Lt,

    /// `>`
    Gt,

    /// `<=`
    Le,

    /// `>=`
    Ge,

    /// `==`
    Eq,

    /// `!=`
    Neq,

    /// `:=`
    Walrus,

    /// `->`
    RArrow,

    /// `fn`
    Fn,

    /// `return`
    Return,

    /// `struct`
    Struct,

    /// `test`
    Test,

    /// end of input
    Eoi,
}
