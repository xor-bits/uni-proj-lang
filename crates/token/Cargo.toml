[package]
name = "token"
version.workspace = true
edition.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
# logos = "0.13"

# only for unit tests
[dev-dependencies]
serde = { version = "1.0", features = ["derive"] }
insta = { version = "1.34", features = ["yaml"] }
