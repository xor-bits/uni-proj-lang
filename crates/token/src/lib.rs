#[cfg(test)]
use serde::{ser::SerializeStruct, Serialize};
use std::ops::Range;

//

macro_rules! some {
    ($v:expr) => {
        if let Some(v) = $v {
            return Some(v);
        }
    };
}

macro_rules! exhaustive {
    ($($tok:ident::$var:ident $({ $($t:tt)* })?,)*) => {{
        use crate::Token::*;

        #[allow(unreachable_code)]
        fn _assert_exhaustive() {
            match unreachable!() {
                $($var { .. } => {})*
            }
        }

        [
            $($tok::$var $({ $($t)* })?),*
        ]
    }};
}

//

pub type Result<T> = core::result::Result<T, Error>;

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy)]
pub enum Error {
    UnexpectedEoi,
    ExtraTokens,
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy)]
pub enum Token {
    /// `struct`
    Struct,

    /// `test`
    Test,

    /// identifiers like `_fn_name`
    Ident,

    /// `;`
    Semi,

    /// `:=`
    Init,

    /// `=`
    Assign,

    /// `(`
    LParen,

    /// `)`
    RParen,

    /// `{`
    Lbrace,

    /// `}`
    Rbrace,

    /// `+`
    Plus,

    /// `-`
    Minus,

    /// `*`
    Asterisk,

    /// `/`
    Slash,

    /// `->`
    Arrow,

    /// `:`
    Type,

    /// `@`
    Extern,

    /// `//`
    LineComment,

    /// integer literal like `0xFFu8`
    LitInt(i128),

    /// string literal like `"test"`
    LitStr,

    /// end of input
    Eoi,
}

impl Token {
    pub const fn enumerate() -> &'static [Self] {
        &exhaustive![
            Token::Struct,
            Token::Test,
            Token::Ident,
            Token::Semi,
            Token::Init,
            Token::Assign,
            Token::LParen,
            Token::RParen,
            Token::Lbrace,
            Token::Rbrace,
            Token::Plus,
            Token::Minus,
            Token::Asterisk,
            Token::Slash,
            Token::Arrow,
            Token::Type,
            Token::Extern,
            Token::LineComment,
            Token::LitInt { 0: 0 },
            Token::LitStr,
            Token::Eoi,
        ]
    }

    pub const fn as_token_str(self) -> Option<&'static str> {
        Some(match self {
            Token::Struct {} => "struct",
            Token::Test => "test",
            Token::Ident => return None,
            Token::Semi => ";",
            Token::Init => ":=",
            Token::Assign => "=",
            Token::LParen => "(",
            Token::RParen => ")",
            Token::Lbrace => "{",
            Token::Rbrace => "}",
            Token::Plus => "+",
            Token::Minus => "-",
            Token::Asterisk => "*",
            Token::Slash => "/",
            Token::Arrow => "->",
            Token::Type => ":",
            Token::Extern => "@",
            Token::LineComment { .. } => "//",
            Token::LitInt { .. } => return None,
            Token::LitStr => return None,
            Token::Eoi => "",
        })
    }
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone)]
pub struct SpannedToken<'a> {
    token: Token,
    span: Span<'a>,
}

impl<'a> SpannedToken<'a> {
    pub fn token(&self) -> Token {
        self.token
    }

    pub fn span(&self) -> Range<usize> {
        self.span.span()
    }

    pub fn str(&self) -> &'a str {
        self.span.str()
    }
}

//

#[derive(Debug, Clone)]
pub struct Span<'a> {
    range: Range<usize>,
    source: &'a str,
}

impl<'a> Span<'a> {
    pub fn span(&self) -> Range<usize> {
        self.range.clone()
    }

    pub fn str(&self) -> &'a str {
        &self.source[self.span()]
    }
}

#[cfg(test)]
impl Serialize for Span<'_> {
    fn serialize<S>(&self, serializer: S) -> core::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut s = serializer.serialize_struct("SpannedToken", 3)?;
        s.serialize_field("range", &format!("{:?}", self.span()))?;
        s.serialize_field("str", &self.str())?;
        s.end()
    }
}

//

#[derive(Debug, Clone)]
pub struct Lexer<'a> {
    source: &'a str,
    at: &'a str,
    eoi: Option<Token>,
    err: Result<()>,
}

impl<'a> Lexer<'a> {
    pub const fn new(source: &'a str) -> Self {
        Self {
            source,
            at: source,
            eoi: Some(Token::Eoi),
            err: Ok(()),
        }
    }

    pub const fn error(&self) -> Result<()> {
        self.err
    }

    fn src_index(&self) -> usize {
        self.at.as_ptr() as usize - self.source.as_ptr() as usize
    }

    fn advance_by(&mut self, len: usize) {
        self.at = &self.at[len..];
    }

    fn advance(&mut self, span: &Span) {
        self.advance_by(span.range.len())
    }

    fn span_to(&self, len: usize) -> Span<'a> {
        let at = self.src_index();
        Span {
            range: at..at + len,
            source: self.source,
        }
    }

    fn spanned_token_to(&mut self, token: Token, len: usize) -> SpannedToken<'a> {
        let span = self.span_to(len);
        self.advance(&span);
        SpannedToken { token, span }
    }

    fn try_match(&mut self, token: Token) -> Option<SpannedToken<'a>> {
        let str = token.as_token_str()?;
        if !self.at.starts_with(str) {
            return None;
        }

        Some(self.spanned_token_to(token, str.len()))
    }

    fn try_match_number(&mut self) -> Option<SpannedToken<'a>> {
        let first = self.at.chars().next()?;

        if !first.is_ascii_digit() {
            return None;
        }

        let mut dot_found = false;
        let term = self.at[1..]
            .find(|c: char| {
                !c.is_ascii_digit()
                    || if dot_found {
                        false
                    } else {
                        dot_found = true;
                        c == '.'
                    }
            })
            .map(|term| term + 1)
            .unwrap_or(self.at.len());

        let token = if dot_found {
            todo!("LitFloat")
        } else {
            let num = self.at[..term].parse::<i128>().unwrap(); // the unwrap should never fail
            Token::LitInt(num)
        };

        Some(self.spanned_token_to(token, term))
    }

    fn try_match_str(&mut self) -> Option<SpannedToken<'a>> {
        if !self.at.starts_with('"') {
            return None;
        }

        let Some(term) = self.at[1..].find('"') else {
            self.err = Err(Error::UnexpectedEoi);
            return self.eoi();
        };

        Some(self.spanned_token_to(Token::LitStr, term + 2))
    }

    fn try_match_ident(&mut self) -> Option<SpannedToken<'a>> {
        let first = self.at.chars().next()?;

        if !(first == '_' || first.is_alphabetic()) {
            return None;
        }

        let term = self.at[1..]
            .find(|c: char| !c.is_alphanumeric())
            .map(|term| term + 1)
            .unwrap_or(self.at.len());

        Some(self.spanned_token_to(Token::Ident, term))
    }

    fn try_match_comment(&mut self) -> Option<SpannedToken<'a>> {
        if !self.at.starts_with("//") {
            return None;
        }

        let term = self.at[2..]
            .lines()
            .next()
            .map(|line| line.len() + 2)
            .unwrap_or(self.at.len());

        Some(self.spanned_token_to(Token::LineComment, term))
    }

    fn eoi(&mut self) -> Option<SpannedToken<'a>> {
        let eoi = self.eoi.take()?;
        Some(self.spanned_token_to(eoi, 0))
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = SpannedToken<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        self.eoi.as_ref()?;
        self.at = self.at.trim();

        // keywords first, so that keywords have a priority over idents
        some!(self.try_match(Token::Struct));
        some!(self.try_match(Token::Test));

        // custom tokens
        some!(self.try_match_str());
        some!(self.try_match_ident());
        some!(self.try_match_comment());
        some!(self.try_match_number());

        // simple tokens in the order of more specific -> less specific
        // length of 2
        some!(self.try_match(Token::Init));
        some!(self.try_match(Token::Arrow));
        // length of 1
        some!(self.try_match(Token::Semi));
        some!(self.try_match(Token::Assign));
        some!(self.try_match(Token::LParen));
        some!(self.try_match(Token::RParen));
        some!(self.try_match(Token::Lbrace));
        some!(self.try_match(Token::Rbrace));
        some!(self.try_match(Token::Type));
        some!(self.try_match(Token::Extern));
        some!(self.try_match(Token::Plus));
        some!(self.try_match(Token::Minus));
        some!(self.try_match(Token::Asterisk));
        some!(self.try_match(Token::Slash));

        if !self.at.is_empty() {
            self.err = Err(Error::ExtraTokens);
        }

        self.eoi()
    }
}

//

#[cfg(test)]
mod tests {
    use crate::{Lexer, Result, SpannedToken, Token};

    use insta::assert_yaml_snapshot;

    fn lex(str: &str) -> (&str, Result<()>, Vec<SpannedToken>) {
        let mut lexer = Lexer::new(str);
        let tokens: Vec<_> = lexer.by_ref().collect();
        (str, lexer.error(), tokens)
    }

    #[test]
    fn lex_test() {
        assert_yaml_snapshot!(lex(
            "in->\"vali\"@->:d synt(ax} {but) t;h:e l=e:=xer should handle it"
        ));

        assert_yaml_snapshot!(lex("_a2"));
        assert_yaml_snapshot!(lex(""));
        assert_yaml_snapshot!(lex("\""));
        assert_yaml_snapshot!(lex("\"a"));
        assert_yaml_snapshot!(lex("\"a\""));

        let all_tokens: String = Token::enumerate()
            .iter()
            .filter_map(|tok| tok.as_token_str())
            .collect();

        assert_yaml_snapshot!(lex(&all_tokens));

        assert_yaml_snapshot!(lex("// comment\ncode"));

        assert_yaml_snapshot!(lex(include_str!("../../../tests/trivial")));
    }
}
