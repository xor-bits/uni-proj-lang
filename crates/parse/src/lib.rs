use std::{fmt, iter::Peekable};

use token::{Lexer, SpannedToken, Token};

//

pub mod ast;

//

pub type Result<T, E = Error> = core::result::Result<T, E>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Error {
    Lexer(token::Error),
    UnexpectedToken(String),
}

impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Error::Lexer(err) => fmt::Display::fmt(err, f),
            Error::UnexpectedToken(tok) => write!(f, "unexpected token `{tok}`"),
        }
    }
}

//

#[derive(Debug, Clone)]
pub enum ParseStream<'a> {
    Lexer(Peekable<Lexer<'a>>),
    Buffer(&'a [SpannedToken<'a>]),
}

impl<'a> ParseStream<'a> {
    pub const fn new(buffer: &'a [SpannedToken<'a>]) -> Self {
        Self::Buffer(buffer)
    }

    pub fn from_lexer(lexer: Lexer<'a>) -> Self {
        Self::Lexer(lexer.peekable())
    }

    pub fn peek(&mut self, token: Token) -> bool {
        Some(token) == self.top().map(|tok| tok.token())
    }

    fn top(&mut self) -> Option<&'_ SpannedToken<'_>> {
        match self {
            ParseStream::Lexer(lexer) => lexer.peek()?.as_ref().ok(),
            ParseStream::Buffer(buffer) => buffer.first(),
        }
    }

    fn top_unexpected(&mut self) -> Error {
        Error::UnexpectedToken(
            self.top()
                .map(|tok| tok.str().to_string())
                .unwrap_or_else(|| "eof".to_string()),
        )
    }

    pub fn parse<T: Parse>(&mut self) -> Result<T> {
        println!("parsing {}", core::any::type_name::<T>());
        T::parse(self)
    }
}

impl<'a> Iterator for ParseStream<'a> {
    type Item = Result<SpannedToken<'a>>;

    fn next(&mut self) -> Option<Self::Item> {
        match self {
            ParseStream::Lexer(lexer) => Some(lexer.next()?.map_err(Error::Lexer)),
            ParseStream::Buffer(buffer) => {
                let (first, others) = buffer.split_first()?;
                *buffer = others;
                Some(Ok(first.clone()))
            }
        }
    }
}

//

pub trait SingleToken: Sized {
    const TOKEN: Token;
    const SELF: Self;

    fn parse_single(tokens: &mut ParseStream) -> Result<Self> {
        if tokens.peek(Self::TOKEN) {
            tokens.next();
            Ok(Self::SELF)
        } else {
            Err(tokens.top_unexpected())
        }
    }
}

//

pub trait Parse: Sized {
    fn parse(tokens: &mut ParseStream) -> Result<Self>;
}

impl<T: Parse> Parse for Box<T> {
    fn parse(tokens: &mut ParseStream) -> Result<Self> {
        Ok(Box::new(tokens.parse()?))
    }
}

impl<T: Parse> Parse for Option<T> {
    fn parse(tokens: &mut ParseStream) -> Result<Self> {
        // cloning the parser is really cheap
        // it would be `Copy` if it wasn't an iterator
        let mut fork = tokens.clone();
        if let Ok(next) = fork.parse() {
            *tokens = fork;
            Ok(Some(next))
        } else {
            Ok(None)
        }
    }
}

impl<T: Parse> Parse for Vec<T> {
    fn parse(tokens: &mut ParseStream) -> Result<Self> {
        let mut result = Vec::new();

        loop {
            // cloning the parser is really cheap
            // it would be `Copy` if it wasn't an iterator
            let mut fork = tokens.clone();
            if let Ok(next) = fork.parse() {
                result.push(next);
                *tokens = fork;
            } else {
                break;
            }
        }

        Ok(result)
    }
}

//

#[cfg(test)]
mod tests {
    use serde::Serialize;
    use token::Lexer;

    use crate::{ast, Parse, ParseStream, Result};

    use insta::assert_yaml_snapshot;

    //

    fn parse<'a, T: Parse + Serialize + 'a>(str: &'a str) -> impl Serialize + 'a {
        let mut stream = ParseStream::from_lexer(Lexer::new(str));
        let ast: Result<ast::Ast<T>, _> = Parse::parse(&mut stream).map_err(|err| err.to_string());
        let extra = stream
            .next()
            .transpose()
            .map(|t| t.map(|t| t.str().to_string()))
            .map_err(|err| err.to_string());

        (str, ast, extra)
    }

    #[test]
    fn parse_test() {
        assert_yaml_snapshot!(parse::<ast::Comma>(","));
    }
}
