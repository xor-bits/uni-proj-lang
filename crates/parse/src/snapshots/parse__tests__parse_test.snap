---
source: crates/parse/src/lib.rs
expression: "parse::<ast::Comma>(\",\")"
---
- ","
- Ok:
    inner: ~
    eoi: ~
- Ok: ~

